sqlitecluster/SQLiteNode.cpp

```diff
@@ -358,27 +386,37 @@ void SQLiteNode::_sendOutstandingTransactions() {
     for (auto& i : transactions) {
         uint64_t id = i.first;
         if (id <= _lastSentTransactionID) {
+            SALERT("Already sent a transaction in committed transaction list");
             continue;
         }
         string& query = get<0>(i.second);
         string& hash = get<1>(i.second);
         uint64_t dbCountAtStart = get<2>(i.second);
-        SData transaction("BEGIN_TRANSACTION");
-        transaction["NewCount"] = to_string(id);
-        transaction["NewHash"] = hash;
-        transaction["leaderSendTime"] = sendTime;
-        transaction["dbCountAtStart"] = to_string(dbCountAtStart);
-        transaction["ID"] = "ASYNC_" + to_string(id);
-        transaction.content = query;
-        _sendToAllPeers(transaction, true); // subscribed only
-        for (auto peer : peerList) {
-            // Clear the response flag from the last transaction
-            (*peer)["TransactionResponse"].clear();
+        string idHeader = to_string(id);
+
+        // If this is marked as "commitOnly", we won't send the BEGIN for it.
+        if (commitOnlyIDs.find(id) == commitOnlyIDs.end()) {
+            // Any commit where we can send a BEGIN and a COMMIT without waiting for acknowledgement is ASYNC.
+            idHeader = "ASYNC_" + idHeader;
+            SData transaction("BEGIN_TRANSACTION");
+            transaction["NewCount"] = to_string(id);
+            transaction["NewHash"] = hash;
+            transaction["leaderSendTime"] = sendTime;
+            transaction["dbCountAtStart"] = to_string(dbCountAtStart);
+            transaction["ID"] = idHeader;
+            transaction.content = query;
+            for (auto peer : peerList) {
+                // Clear the response flag from the last transaction
+                (*peer)["TransactionResponse"].clear();
+            }
+            _sendToAllPeers(transaction, true); // subscribed only
+        } else {
+            SINFO("Sending COMMIT for QUORUM transaction " << idHeader << " to followers");
         }
         SData commit("COMMIT_TRANSACTION");
-        commit["ID"] = transaction["ID"];
-        commit["CommitCount"] = transaction["NewCount"];
-        commit["Hash"] = hash;
+        commit["ID"] = idHeader;
+        commit["NewCount"] = to_string(id);
```

Comment:
isn't this the same as idHeader now?

Reply:
Yes. I don't know if it's clearer that this one definitely doesn't contain "ASYNC_" or not. Could go either way here.

Reply:
On second thought, no, it's only the same for quorum transactions, it's different for ASYNC transactions.